privacyidea/lib/crypto.py

```diff
@@ -171,115 +136,91 @@ def _clearKey_(self, preserve=False):
-#         error = "Error during decoding activationcode %r: %r" % (acode, exx)
-#         log.error(error)
-#         raise Exception(error)
-#
-#     if checksum is True:
-#         checkCode = str(activationcode[-2:])
-#         veriCode = str(check(bcode)[-2:])
-#         if checkCode != veriCode:
-#             raise Exception('[crypt:kdf2] activation code checksum error!! '
-#                             ' [%s]%s:%s' % (acode, veriCode, checkCode))
-#
-#     activ = binascii.hexlify(bcode)
-#     passphrase = u'' + sharesecret + activ + nonce[:-salt_len]
-#     #keyStream = PBKDF2(binascii.unhexlify(passphrase),
-#     #                   bSalt, iterations=iterations,
-#     #                   digestmodule=digestmodule)
-#     #key = keyStream.read(length)
-#     key = pbkdf2_sha256(binascii.unhexlify(passphrase),
-#                         salt=bSalt, rounds=iterations)
-#     return key
+def _to_bytes(s):
+    """
+    Convert string to bytes if it isn't already
+    :param s: string to convert
+    :type s: str, bytes, unicode
+    :return: the converted string
+    :rtype: bytes
+    """
+    if isinstance(s, six.text_type):
+        return s.encode('utf8')
+    return s
+
+
+def _to_unicode(s):
+    """
+    decode a byte string to unicode using utf8
+    :param s: string to decode to utf8
+    :type s: bytes
+    :return: the utf-8 decoded string s
+    :rtype: str
+    """
+    if isinstance(s, bytes):
+        return s.decode('utf8')
+    return s
+
+
+def hexlify_and_unicode(s):
+    """
+
+    :param s: string to hexlify
+    :type s: bytes
+    :return: hexlified string converted to unicode
+    :rtype: unicode
+    """
+    return _to_unicode(binascii.hexlify(s))
 
 
 @log_with(log, log_entry=False, log_exit=False)
 def hash(val, seed, algo=None):
+    """
+
+    :param val: value to hash
+    :type val: bytes or str or unicode
+    :param seed: seed for the hash
+    :type seed: bytes, str, unicode
+    :param algo:
+    :return: the hexlified hash value calculated from hash and seed
+    :rtype: str
+    """
     log.debug('hash()')
     m = sha256()
-    m.update(val.encode('utf-8'))
-    m.update(seed)
-    return m.digest()
+    m.update(_to_bytes(val))
+    m.update(_to_bytes(seed))
+    return hexlify_and_unicode(m.digest())
 
 
-def hash_with_pepper(password, rounds=10023, salt_size=10):
+def hash_with_pepper(password):
     """
     Hash function to hash with salt and pepper. The pepper is read from
     "PI_PEPPER" from pi.cfg.
 
     Is used with admins and passwordReset
 
-    :return: Hash string
+    :param password: the password to hash
+    :type password: str
+    :return: hashed password string
+    :rtype: str
     """
     key = get_app_config_value("PI_PEPPER", "missing")
-    pw_dig = passlib.hash.pbkdf2_sha512.encrypt(key + password, rounds=rounds,
-                                                salt_size=salt_size)
+    pw_dig = passlib.hash.pbkdf2_sha512.hash(key + password)
```

Comment:
The documentation recommends keeping the default values. I am not sure what we would gain (in our case even reducing the number of `rounds` and the `salt_size` compared to the default).
I would prefer sticking with the default since 'they know best'...

Reply:
You are probably right. I also think the parameters where not used externally. We added them to be able to allow someone who "thinks he knows better", to more easily adapt this.

Reply:
:+1: